### DISCUSSION TITLE
File watcher limit exceeded when using histoire build

### DISCUSSION THREAD

**Discussion by anoack93** [histoire-dev/histoire#291] [Q&A]
When using histoire@0.10.7 (vite 3.1.0) and executing `histoire build` in our GitLab CI/CD pipeline the job fails due to a file system limit which is being reached. I use histoire with nuxt and also run nuxt build in advance. I don't get why histoire needs file system watchers, when running the production build. Is that a known problem?

```
histoire build
node:internal/errors:465
    ErrorCaptureStackTrace(err);
    ^
Error: ENOSPC: System limit for number of file watchers reached, watch '/builds/foo-bar/starter/.npm/_cacache/content-v2/sha512/64/36/...'
    at FSWatcher.<computed> (node:internal/fs/watchers:244:19)
    at Object.watch (node:fs:2264:34)
    at createFsWatchInstance (file:///builds/foo-bar/starter/node_modules/vite/dist/node/chunks/dep-665b0112.js:49738:17)
    at setFsWatchListener (file:///builds/foo-bar/starter/node_modules/vite/dist/node/chunks/dep-665b0112.js:49785:15)
    at NodeFsHandler$1._watchWithNodeFs (file:///builds/foo-bar/starter/node_modules/vite/dist/node/chunks/dep-665b0112.js:49940:14)
    at NodeFsHandler$1._handleFile (file:///builds/foo-bar/starter/node_modules/vite/dist/node/chunks/dep-665b0112.js:50004:23)
    at NodeFsHandler$1._addToNodeFs (file:///builds/foo-bar/starter/node_modules/vite/dist/node/chunks/dep-665b0112.js:50246:21)
Emitted 'error' event on FSWatcher instance at:
    at FSWatcher._handleError (file:///builds/foo-bar/starter/node_modules/vite/dist/node/chunks/dep-665b0112.js:51434:10)
    at NodeFsHandler$1._addToNodeFs (file:///builds/foo-bar/starter/node_modules/vite/dist/node/chunks/dep-665b0112.js:50254:18) {
  errno: -28,
  syscall: 'watch',
  code: 'ENOSPC',
  path: '/builds/foo-bar/starter/.npm/_cacache/content-v2/sha512/64/36/...',
  filename: '/builds/foo-bar/starter/.npm/_cacache/content-v2/sha512/64/36/...'
}
```

**Comment by Akryum** [comment]
I guess this is caused by the watcher that is used to scan for markdown files:
https://github.com/histoire-dev/histoire/blob/b57aad827572cc3ac2bff8ce1b955ec61c2bee48/packages/histoire/src/node/build.ts#L45

**Reply by anoack93** [reply]
I tried to find the root cause of this problem, but I was not very successful yet.

I found out the problem occurs only in the CI/CD pipeline because of the way we cache node modules.

We are using `npm i --cache .npm --prefer-offline` to create the npm cache directly inside the working directory, where also the nuxt 3 application and histoire lives. This is needed for GitLab CI caching. This causes the file watcher problem. I could verify that the problem does not occur if the caching directory is not changed.
So it seems like some scanning mechanism in histoire build causes too many files in the .npm directory being watched.
Thats why I added the .npm directory to storyIgnored

```
  storyIgnored: [
    '**/node_modules/**',
    '**/dist/**',
    '**/.npm/**',
  ],
```

but this does not fix the problem either, so I don't know whether createMarkdownFilesWatcher causes the problem, as it should use the storyIgnored config too.

But in general it would probably be more reliable to not use file watchers when indexing files in build mode.

**Reply by Akryum** [reply]
> But in general it would probably be more reliable to not use file watchers when indexing files in build mode.

Agreed

**Reply by damianidczak** [reply]
Hi,

I have the same problem after upgrade my Library (not histoire). I removed from build 'umd'  format and I set only 'es' format.

**Reply by freakzlike** [reply]
The issue still seems to exist and I am unable to build within the CI/CD Pipeline. Even if I don't have any `*.story.md` in my repository. I just wonder which files are being watch then?
https://github.com/histoire-dev/histoire/blob/b2bc00983726fa49eafd29386abf6823ebbdd942/packages/histoire/src/node/markdown.ts#L138-L141

**Reply by anoack93** [reply]
@freakzlike I was able to resolve this issue by increasing the file watcher limit right before running the build in the pipeline

```
echo fs.inotify.max_user_watches=524288 | tee -a /etc/sysctl.conf && sysctl -p
```

the actual way to change the limit depends on the base image your are using to run your CI/CD jobs
